gcp/gcp.go

```diff
@@ -0,0 +1,108 @@
+// GCPのライブラリを使った処理を集めるパッケージ
+package gcp
+
+import (
+	"context"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+
+	"cloud.google.com/go/storage"
+	"github.com/golang/snappy"
+	"google.golang.org/api/option"
+)
+
+// GCPを扱う構造体の定義
+type Handler struct {
+	LocalPath    string
+	GcpKey       string
+	ProjectId    string
+	StorageClass string
+	Duration     int64
+}
+
+func (env *Handler) CreateClient() (*storage.Client, error) {
+	// jsonで渡された鍵のサービスアカウントに紐づけられたクライアントを建てる
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx, option.WithCredentialsFile(env.GcpKey))
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("Successfully set a client in \"%s\"", env.ProjectId)
+	return client, err
+}
```

Comment:
GCPを扱う構造体は、いわゆるクライアントだと思えばよいです。
つまり、既存のクライアントをラップして使いやすいクライアントを作るわけです。
```suggestion
// GCPを扱う構造体の定義
// LocalPathはGCPには関係なくてクライアントを使うときに必要な値なのでここでは使わない。
type Client struct {
	ProjectID    string
	StorageClass string
	Duration     int64
	storage.Client
}

func NewClient(key, projectID, storageClass string, duration int64) (*Client, error) {
	// jsonで渡された鍵のサービスアカウントに紐づけられたクライアントを建てる
	ctx := context.Background()
	client, err := storage.NewClient(ctx, option.WithCredentialsFile(key))
	if err != nil {
		return nil, err
	}

	log.Printf("Successfully set a client in \"%s\"", projectID)
	return &Client{ projectID, storageClass, duration, client }, err
}
```
それで、↑にメソッドを生やせばよいです。
と、ここまで説明してなんですが個人的にはこの大きさのプログラムならパッケージを切る必要もなくてファイルを分けて関数に切り出すぐらいでいいと思ってます。
それで環境変数は`main.go`のグローバル変数に持って関数に渡すようにすれば良いんじゃないかな？

Reply:
環境変数は`config`パッケージに`Load`みたいな関数用意して構造体返すようにするのがよさそうだとぼくは思います

Reply:
`確かに！GCPではクライアントが該当しますね。気づかなかった。`
んでもって結局感がすごい… コードの分け方とか変数の持ち方・定義の仕方・引数の形とかって人それぞれでやり方ありますよね… わかってはいた
う～ん… 環境設定は構造体と読み込み関数を別ファイルで用意してそいつを引数で渡す感じでまとめようと思います。パッケージまとめるかどうかは今回くらいだとどっちでもいい感じなのかな？プロジェクトのサイズが大きければ分割した方が良いんですね。今回まとめるかはちょっと考えます。

今回とりあえず、最低限
・環境変数は一個の関数でまとめて読み込もう
・メソッドを生やすときは概念的に大きな型にして生やそう
・引数を渡すときは抽象性をできるだけ高めよう
は全体的にマナーとして守ったほうがいいんじゃないか、と思いましたが、これ以外に今まで挙げられた問題点の中で守った方が良い"一般的"な(=人によらない)マナーってありますかね？(Goの基本的な文法以外の理解しやすいコードを書くという面で)

Reply:
環境変数を構造体として取りまわすならパッケージ切ったほうがよくて、別にそうしないならmainで読み込んじゃっていいと思ってます
今のコードは構造体として取りまわしてるのでパッケージ切ったほうがいいって言いましたが、大きさ的には別に切り分ける必要はないと思ってます

こういう(=設計とかアーキテクチャって言われる話)のは今じゃなくて将来に焦点を当てて考えるもので、個人間での好みというよりはそのコードの未来に何を見てるかって感じがします

Reply:
「構造体で取り回すなら切った方がいい」の理由がピンときてないのですがどうしてでしょう？配列的な構造体の使い方だったらいいとかそういうのってあります？
アーキテクチャ、聞いたことはあります！覚えておかないと… まだ「目の前の実装を綺麗にまとめる」までしか考えられてないので未来でどう使われるか考えられるようにならないとなと思います

Reply:
> 「構造体で取り回すなら切った方がいい」の理由

これは単純に構造体で取り回すようにするとその関連のコードが増えるからですね

Reply:
なるほど…？これは多分もっと複雑なコードにならないとありがたみがわからないやつかもしれない…
とりあえず今回はmainパッケージに素でグローバル変数として環境変数置きました